core/client/src/main/java/alluxio/client/block/AlluxioBlockStore.java

```diff
@@ -40,20 +44,37 @@
 public final class AlluxioBlockStore {
   private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
 
-  private static AlluxioBlockStore sClient = null;
+  // Only one block store will be kept for each master.
+  private static final Map<InetSocketAddress, AlluxioBlockStore> CACHED_BLOCK_STORES =
+      new ConcurrentHashMap<>();
 
   /**
+   * Gets a block store using the master address got from config.
+   *
    * @return the Alluxio block store instance
    */
   public static synchronized AlluxioBlockStore get() {
-    if (sClient == null) {
-      sClient =
-          new AlluxioBlockStore(BlockStoreContext.INSTANCE, NetworkAddressUtils.getLocalHostName());
+    return get(ClientContext.getMasterAddress());
+  }
+
+  /**
+   * Gets a block store with the specified master address from the cache if it's created before.
+   * Otherwise creates a new one and puts it in the cache.
+   *
+   * @param masterAddress the master's address
+   * @return the block store created or cached before
+   */
+  public static synchronized AlluxioBlockStore get(InetSocketAddress masterAddress) {
+    AlluxioBlockStore store = CACHED_BLOCK_STORES.get(masterAddress);
+    if (store == null) {
+      store = new AlluxioBlockStore(BlockStoreContext.get(masterAddress),
+          NetworkAddressUtils.getLocalHostName());
+      CACHED_BLOCK_STORES.put(masterAddress, store);
     }
-    return sClient;
+    return store;
   }
 
-  private final BlockStoreContext mContext;
+  private BlockStoreContext mContext;
```

Comment:
not introduced by this PR, but normally fields go at the top and I would be a lot happier if these could move up ;)